### PR TITLE
Shovel/Shield Knight

### PR DESCRIPTION
**Pull request by Mach131**
probably extremely messy but all the tests are passing

### CODE REVIEW

File: scenes/game/local_game.gd [hunk 1]
```diff
@@ -2285,6 +2519,10 @@ func handle_strike_effect(card_id :int, effect, performing_player : Player):
 			performing_player.strike_stat_boosts.ignore_guard = true
 		"ignore_push_and_pull":
 			performing_player.strike_stat_boosts.ignore_push_and_pull = true
+		"ignore_push_and_pull_boost":
+			performing_player.ignore_push_and_pull = true
```
Comment (daniel-k-taylor) [inline scenes/game/local_game.gd:2523]: maybe we should call this one "ignore_push_and_pull_boost_bonus" just to be super clear and not confuse with the other one
Or maybe passive_bonus? passive_boost?

File: scenes/game/local_game.gd [hunk 2]
```diff
@@ -3156,12 +3450,19 @@ func do_hit_response_effects(offense_player : Player, defense_player : Player, i
...
 	var events = []
 	active_strike.strike_state = next_state
-	if not offense_player.strike_stat_boosts.ignore_armor and defense_player.strike_stat_boosts.when_hit_force_for_armor:
-		change_game_state(Enums.GameState.GameState_PlayerDecision)
-		decision_info.player = defense_player.my_id
-		decision_info.type = Enums.DecisionType.DecisionType_ForceForArmor
-		decision_info.choice_card_id = active_strike.get_player_card(defense_player).id
-		events += [create_event(Enums.EventType.EventType_Strike_ForceForArmor, defense_player.my_id, incoming_damage)]
+	if not offense_player.strike_stat_boosts.ignore_armor:
+		# Currently assumes these will be armor-related; probably breaks if choices get involved
```
Comment (daniel-k-taylor) [inline scenes/game/local_game.gd:3454]: almost certainly breaks if choices get involved and block is also involved.
If someone has multiple when_hit that are choices or need to be ordered and interact with block, we'll have to add the same kind of effect list stuff as other timing windows.
For now, this should be fine though.

File: scenes/game/local_game.gd [hunk 3]
```diff
@@ -3156,12 +3450,19 @@ func do_hit_response_effects(offense_player : Player, defense_player : Player, i
...
+		if active_strike.initiator == defense_player:
+			events += do_effects_for_timing("when_hit", defense_player, active_strike.initiator_card, next_state)
```
Comment (daniel-k-taylor) [inline scenes/game/local_game.gd:3456]: I'm not a huge fan of passing in the card when it doesn't matter, but I see you're doing that as just a way to get the character effects....
Maybe just do what reshuffle_discard does?
var effects = get_character_effects_at_timing("on_reshuffle")
for effect in effects:
    events += parent.do_effect_if_condition_met(self, -1, effect, local_conditions)

I don't want to read this later and accidentally think it does more than it does.

File: scenes/game/local_game.gd [hunk 4]
```diff
@@ -1122,14 +1145,38 @@ class Player:
...
+			on_buddy_space = true
+		else:
+			if on_buddy_space:
+				handle_on_buddy_boosts(false)
+			on_buddy_space = false
+
 	func move_to(new_arena_location):
 		var events = []
+
+		if cannot_move:
```
[the remark this answers is not in the document]
Comment (daniel-k-taylor) [reply scenes/game/local_game.gd:1161]: Okay, I saw later why this was here from the swap places effect.
Should be fine.

**Review by daniel-k-taylor — CHANGES_REQUESTED** [review]

File: scenes/game/ai/ai_policy_rules.gd [hunk 5]
```diff
@@ -106,13 +110,20 @@ func can_card_hit(card_id : int, ex_card_id : int, ai_game_state : AIPlayer.AIGa
...
+			var result = get_locations_after_effect(effect, my_location, opponent_location, buddy_location)
 			my_location = result['my_location']
 			opponent_location = result['opponent_location']
+			buddy_location = result['buddy_location']
+		if effect['effect_type'] == "calculate_range_from_buddy":
+			from_buddy = true
 
 	var distance_after_effects = abs(my_location - opponent_location)
+	if from_buddy:
+		distance_after_effects = abs(buddy_location - opponent_location)
```
[the remark this answers is not in the document]
Comment (Mach131) [reply scenes/game/ai/ai_policy_rules.gd:126]: it be like that; at least it's less likely to throw out shield knight attacks when she's nowhere near you

File: scenes/game/local_game.gd [hunk 6]
```diff
@@ -712,13 +733,15 @@ class Player:
...
 		var events = []
 		buddy_location = new_location
+		update_on_buddy_space()
 		events += [parent.create_event(Enums.EventType.EventType_PlaceBuddy, my_id, buddy_location)]
 		return events
 
 	func remove_buddy():
 		var events = []
-		if not do_not_cleanup_buddy_this_turn:
+		if not do_not_cleanup_buddy_this_turn and not buddy_always_on_field:
```
[the remark this answers is not in the document]
Comment (Mach131) [reply scenes/game/local_game.gd:742]: are there any characters like that? will prob just make it starting_offset to be safe but might be able to get away with a bool

File: scenes/game/local_game.gd (hunk 4 above)
[the remark this answers is not in the document]
Comment (Mach131) [reply scenes/game/local_game.gd:1161]: yee hopefully shouldn't break regular movement

File: scenes/game/local_game.gd [hunk 7]
```diff
@@ -1139,6 +1186,22 @@ class Player:
 		if cannot_move:
 			events += [parent.create_event(Enums.EventType.EventType_BlockMovement, my_id, 0)]
 			return events
+
+		# avoiding double close effects
+		if not resolving_close_effect:
+			var effects = get_character_effects_at_timing("on_advance_or_close")
```
[the remark this answers is not in the document]
Comment (Mach131) [reply scenes/game/local_game.gd:1192]: I think I got it, just had to do a bit of extra stuff to make sure i'm not losing the and effects

File: scenes/game/game.gd [hunk 8]
```diff
@@ -1026,20 +1026,21 @@ func _on_name_opponent_card_begin(event):
...
 		select_card_require_min = 1
 		select_card_require_max = 1
-		var cancel_allowed = false
+		var cancel_allowed = can_name_fake_card
 		popout_instruction_info = {
 			"popout_type": CardPopoutType.CardPopoutType_ReferenceOpponent,
 			"instruction_text": "Name an opponent card.",
 			"ok_text": "OK",
-			"cancel_text": "",
+			"cancel_text": "Name Exodia",
```
Comment (daniel-k-taylor) [inline scenes/game/game.gd:1041]: funny, but maybe just "Name Nothing - Reveal Hand" is better

Comment (daniel-k-taylor) [reply scenes/game/game.gd:1041]: I guess you could leave it Exodia, but just add the Reveal Hand part to be clear what is going on for people who don't get it

**Review by daniel-k-taylor — APPROVED** [review]

**Review by daniel-k-taylor — APPROVED** [review]

**Pull request merged**